### PR TITLE
prevent allied units from autoattack building which is being captured…

### LINKED ISSUE #6170: Allied units shouldn't attack a building which is being captured [$20]
Allied units shouldn't attack a building if an engineer is in the process of capturing it.

<bountysource-plugin>
---
There is a **[$20 open bounty](https://www.bountysource.com/issues/3478400-allied-units-shouldn-t-attack-a-building-which-is-being-captured?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Arcitectus**
…. resolves #6170.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/ExternalCapturable.cs [hunk 1]
```diff
@@ -91,5 +91,15 @@ public void Tick(Actor self)
...
 		}
+
+		public bool PreventAutoTarget(Actor self, Actor attacker)
+		{
+			if (Captor == null)
+			{
+				return false;
+			}
+
+			return attacker.AppearsFriendlyTo(Captor);
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/ExternalCapturable.cs]: Could you change this entire block to `return Captor != null && attacker.AppearsFriendlyTo(Captor);`, please?

**Comment by Phrohdoh** [comment]
:+1: after that change.
This works and is pretty neat.

**Comment by GraionDilach** [comment]
I request a tag to make this optional.

File: OpenRA.Mods.Common/Traits/AutoTarget.cs [hunk 2]
```diff
@@ -167,15 +167,17 @@ void Attack(Actor self, Actor targetActor, bool allowMove)
 
 		Actor ChooseTarget(Actor self, WDist range, bool allowMove)
 		{
-			var inRange = self.World.FindActorsInCircle(self.CenterPosition, range)
-				.Where(a => !a.Info.HasTraitInfo<AutoTargetIgnoreInfo>());
+			var inRange = self.World.FindActorsInCircle(self.CenterPosition, range);
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/AutoTarget.cs]: This isn't used anywhere else, so why split it out?

File: OpenRA.Mods.Common/Traits/AutoTarget.cs [hunk 3]
```diff
@@ -168,7 +168,8 @@ void Attack(Actor self, Actor targetActor, bool allowMove)
 		Actor ChooseTarget(Actor self, WDist range, bool allowMove)
 		{
 			var inRange = self.World.FindActorsInCircle(self.CenterPosition, range)
-				.Where(a => !a.Info.HasTraitInfo<AutoTargetIgnoreInfo>());
+				.Where(a =>
+				!a.TraitsImplementing<IPreventsAutoTarget>().Any(t => t.PreventAutoTarget(a, self)));
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/AutoTarget.cs]: Our style convention is to indent cases like this 2nd line with another tab (to make it more visible that the line belongs to the `Where` above).

**Comment by Phrohdoh** [comment]
My comment was read before being updated.
The original was this:

Could you add a bool AutoTargetIgnores to the ExternalCapturableInfo class so this can be (un)set per-actor type then change the entire block to `return Info.AutoTargetIgnores && Captor != null && attacker.AppearsFriendlyTo(Captor);`, please?

**Comment by penev92** [comment]
I'd like to hear opinions on two things:
IPreventsAutoTarget -> IPreventAutoTarget
and
bool PreventAutoTarget() -> bool IsPreventingAutoTarget()

**Comment by GraionDilach** [comment]
IPreventAutoTarget and PreventAutoTarget imo.

**Comment by Phrohdoh** [comment]
@penev92 I agree on both naming changes, and have another request now that the info field has been added.

File: OpenRA.Mods.Common/Traits/ExternalCapturable.cs [hunk 4]
```diff
@@ -21,6 +21,7 @@ public class ExternalCapturableInfo : ITraitInfo
 		public readonly bool AllowAllies = false;
 		public readonly bool AllowNeutral = true;
 		public readonly bool AllowEnemies = true;
+		public readonly bool AutoTargetIgnores = false;
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/ExternalCapturable.cs]: Please add a `desc` (like the one below) to this trait property describing the circumstance that this is tied to.

Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/ExternalCapturable.cs]: Also a newline between `AllowEnemies` and the `desc` imo.

**Comment by Arcitectus** [comment]
I applied the suggested changes, default is now the old behaviour and the change can be activated by setting `AutoTargetIgnores`.

**Comment by GraionDilach** [comment]
:+1:

File: OpenRA.Mods.Common/Traits/ExternalCapturable.cs [hunk 5]
```diff
@@ -24,6 +24,9 @@ public class ExternalCapturableInfo : ITraitInfo
 		[Desc("Seconds it takes to change the owner.", "You might want to add a ExternalCapturableBar: trait, too.")]
 		public readonly int CaptureCompleteTime = 15;
 
+		[Desc("Whether autotargeting ignores this actor while it is being captured by an ally.")]
+		public readonly bool AutoTargetIgnores = false;
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/ExternalCapturable.cs]: I think you can safely set this to `true`. At least for the monster tank madness mission.

Comment (Arcitectus) [inline OpenRA.Mods.Common/Traits/ExternalCapturable.cs]: @Mailaender 
> I think you can safely set this to true. At least for the Monster Tank Madness mission.

I am not sure what that means in relation to the PR.
I thought that this was the wrong place to implement a setting for a specific mission. Which led me to believe that your comment was just meant as a note to users of the field. Is that correct?
Or do you think changes should be made to the PR?

Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/ExternalCapturable.cs]: I think changing individual missions is out of scope for this PR.

**Comment by penev92** [comment]
Needs a rebase.

File: OpenRA.Mods.Common/Traits/ExternalCapturable.cs (hunk 5 above)
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/ExternalCapturable.cs]: I'm with @Mailaender here. If you don't set this to true, this doesn't really fix #6170 (except you add some yaml to the default mods), as the behaviour ingame won't change.

**Comment by Phrohdoh** [comment]
This was silently rebased.

**Comment by Phrohdoh** [comment]
I suspect something is missing from this, I'll investigate and post my findings.

**Comment by Phrohdoh** [comment]
~delete~

**Comment by Arcitectus** [comment]
@Phrohdoh 
I changed the names of interface and its method to match the examples you cited.

File: OpenRA.Mods.Common/Traits/ExternalCapturable.cs [hunk 6]
```diff
@@ -24,6 +24,9 @@ public class ExternalCapturableInfo : ITraitInfo
 		[Desc("Seconds it takes to change the owner.", "You might want to add a ExternalCapturableBar: trait, too.")]
 		public readonly int CaptureCompleteTime = 15;
 
+		[Desc("Whether autotargeting ignores this actor while it is being captured by an ally.")]
+		public readonly bool AutoTargetIgnores = true;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/ExternalCapturable.cs]: Can we change this to `PreventsAutoTarget` to match the other recent changes?

**Comment by Phrohdoh** [comment]
Thanks Arcitectus! Sorry for all of the back and forth.

Just one more thing to change then this should be on its way. :)

**Comment by Arcitectus** [comment]
> Just one more thing to change

done.

**Comment by obrakmann** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/740cd9a2afd1905e25ad03a8121503aab2a17426).

@Arcitectus, don't forget to collect your bounty!

**Comment by r34ch** [comment]
@Arcitectus Currently on bleed I have found two issues;

1. If a unit is attacking a building and then a friendly engineer starts capturing it, that unit will not stop attacking until ordered to move; with the exception of

2. turreted units will continue to attack the building being captured even when moved, including moving out of range and back into range.

Both attacking issues can be stopped by manually hitting the S hotkey.

**Comment by Arcitectus** [comment]
@r34ch I will look for a way to invalidate the targets for those scenarios.

**Comment by GraionDilach** [comment]
2) is pretty much beyond this one's scope a lot, forceattack terrain works that way and prolly a lot of stuff the recent targeting PRs are touching.

**Pull request merged**